Modules/Sources/JetpackStatsWidgetsCore/WidgetDataCacheReader.swift

```diff
@@ -0,0 +1,30 @@
+public protocol WidgetDataCacheReader {
+    func widgetData<T: HomeWidgetData>(for siteID: String) -> T?
+    func widgetData<T: HomeWidgetData>() -> [T]?
+}
+
+public extension WidgetDataCacheReader {
+
+    func widgetData<T: HomeWidgetData>(
+        forSiteIdentifier identifier: String?,
+        defaultSiteID: Int?,
+        userLoggedIn: Bool
+    ) -> Result<T, WidgetDataReadError> {
+        if let selectedSite = identifier, let widgetData: T = widgetData(for: selectedSite) {
+            return .success(widgetData)
+        } else if let defaultSiteID, let widgetData: T = widgetData(for: String(defaultSiteID)) {
+            return .success(widgetData)
+        } else {
+            if userLoggedIn {
+                /// In rare cases there could be no default site and no defaultSiteId set
+                if let firstSiteData: T = widgetData()?.sorted(by: { $0.siteID < $1.siteID }).first {
+                    return .success(firstSiteData)
+                } else {
+                    return .failure(.noSite)
+                }
+            } else {
+                return .failure(.loggedOut)
+            }
+        }
+    }
```

Comment:
This is the core change that makes everything else happen. If you look at it together with the diff in `WidgetDataReader` below, you'll see how this is the exact logic that was there before, with the difference that it doesn't access higher-level objects such as `UserDefaults` but simply expects all the raw information to be given to it as inputs.

The advantage of this approach is that it becomes decoupled from the app or Foundation so we can test it in isolation. It's also handy because it splits the non-trivial decision logic with the logic to gather the inputs for the decision.